solver/scheduler.go

```diff
@@ -351,7 +352,7 @@ type pipeFactory struct {
 func (pf *pipeFactory) NewInputRequest(ee Edge, req *edgeRequest) pipe.Receiver {
 	target := pf.s.ef.getEdge(ee)
 	if target == nil {
-		panic("failed to get edge") // TODO: return errored pipe
+		return pipe.NewErroredPipe(req, fmt.Errorf("failed to get edge"))
```

Comment:
would it be easier to call `pf.NewFuncRequest(func() {return errors.Errorf()})` here?

Reply:
If I got it right, this would only fail the sender pipe and not the receiver, which we are looking for in `NewInputRequest()`

Reply:
not sure I understand.

Reply:
Said differently, `pf.NewFuncRequest(func() {return errors.Errorf()})` doesn't return an errored `pipe.Receiver`. If you call 	`.Status().Err` on it, you get `nil`

Reply:
> If you call .Status().Err on it, you get nil

After you call, `Receive` you still get nil?

Reply:
Yes. Maybe I'm not getting something right :)
```
func TestPanic(t *testing.T) {
	s := NewSolver(SolverOpt{
		ResolveOpFunc: testOpResolver,
	})
	defer s.Close()

	e := edge{}
	pf := &pipeFactory{s: s.s, e: &e}
	receivedPipe := pf.NewFuncRequest(func(_ context.Context) (interface{}, error) { return nil, errors.Errorf("failed to get edge") })
	receivedPipe.Receive()
	require.Error(t, receivedPipe.Status().Err)
}
```

We only need an errored pipe in this one place and therefore I have nothing against dropping the special purpose errored pipe structure I introduce. On the other hand, it's quite simple to understand.

Reply:
The event is already captured by the solver internally.

```

func TestPanic(t *testing.T) {
	f := func(_ context.Context) (interface{}, error) {
		return nil, errors.Errorf("failed to get edge")
	}
	p, start := pipe.NewWithFunction(f)
	go start()

	for {
		if !p.Receiver.Receive() {
			time.Sleep(time.Millisecond) // in real code this is achieved with signal
			continue
		}
		require.Error(t, p.Receiver.Status().Err)
		break
	}
}
```

Reply:
In your example you can replace `receivedPipe.Receive()` with `time.Sleep(time.Second)`. Function runs async (and  `Receive()` does not block). You can't get the actual function completion time or wait for `Receive()` as these are already captured internally in schedulers loop thread.

Reply:
Thanks for the explanation !